Os.exec win32

coreyapowell opened the pull request:
I think I have this figured out, minimized changes for win32 functionality and testing of os.exec.
 someone should test on a real linux machine .. and you're doing mac?
I had to change several #ifdef (__wasi__) lines. I wasn't expecting to just change the codebase. I'm sorry about going back and forth, but someone changed permissions while I was in the middle of uploading files and it's created a huge mess that I can't seem to fix!
downloaded compiled and tests run with MING64/Clang64 Msys2, Debian on WSL, MSVC 2022, Clang for MSVC

coreyapowell committed:
Add files via upload

coreyapowell committed:
Add files via upload

coreyapowell committed:
Add files via upload

coreyapowell committed:
Add files via upload

coreyapowell committed:
Update test_std.js

I put this back like it was. 
Changes will be split into smaller parts and I've gotten confused when permissions went back and forth

coreyapowell committed:
Merge pull request #2 from coreyapowell/backto_original

Update test_std.js

coreyapowell committed:
Delete test_std.js

getting back to original codebase

coreyapowell committed:
adding test script for win32 os.exec

tests on win32:
os.exec
os.pipe (fd version)
os.kill
os.watchpid
 
maintains test on linux:
os.exec
os.kill
os.waitpid
os.watchpid
os.pipe

coreyapowell committed:
os_exec

adds os_exec, os_pipe, build_ms_envp, os_watchpid (2 versions) os_kill

coreyapowell committed:
Update and rename test_win_exec.js to test_os_exec.js

just modified comments and changed name. 
I do plan to test watchpid on non-win32.

coreyapowell committed:
Update test_os_exec.js

managed to also verify execution of .exe, return values of exe, and stderr handle.  verifies watchpid in non-win32 for compatible scripting. the only non-compatible issue left I see is that getline() pulls from windows needing \r to be trimmed from <eol>

coreyapowell committed:
Update quickjs-libc.c

jscheck issue fixed.
I did move the functions around to minimize code differences. env in win32 is one big zero-delimited buffer; I now double its size and copy on overflow.

coreyapowell committed:
Update quickjs-libc.c

it's semantics. I just tested it. 
my compiler is not liking size_t addition to char* and for some reason even code editor seems to think it's typecasting crazy. so I'm comparing references to keep it clean enough looking.

coreyapowell committed:
Update quickjs-libc.c

needed to free the memory when I re-cache.

coreyapowell committed:
Update quickjs-libc.c

tested again ...
I had comments and printf functions to remove. I just keep reviewing and expanding my test environment.

bnoordhuis reviewed (commented):
There are a bunch of style issues, like inconsistent whitespace, non-idiomatic (for this project) if/else style, etc. Can you fix those first?

Rule of thumb: new code should look like the existing code that surrounds it.

coreyapowell commented:
Sure. 
I'm trying to understand how you guys collaborate as independent users.  I want whoever is handling the windows portion to at least have a heads up. It's as 'drop in replacement' as I could manage. I just expected someone might have some feedback.

I don't want to trample over everyone else's projects. In fact, I believe Mr Bellard likely ran into the same issues (that windows doesn't play nice with FDs and PIDs) and just skipped all of this. The quickjs-ng team added worker pipes for win32 (in quickjs-libc.c), but they use HND types in a piece of code that users won't see in the script. The quickjs documentation, the current std and os module, the c standard, and all of linux use and prefer PIDs and FDs. I haven't had a chance to test things like passing the FD down a pipe to another .exe or to a worker thread. Passing out HNDs in script would guarantee that flexibility, but it would be a documentation nightmare. So this works, it does what I need it to do as a scripting tool, and it function simply enough. 

I've created one replacement function that can be used in linux and windows without testing the environment, but the user needs to understand what that function does. Windows doesn't have waitpid. Do I just leave it in the tests.js folder and let them figure it out? Leave my comments above the c code and test.js code? Or put a comment in the documentation section for watchpid next to waitpid? Is everyone okay with the way I split those functions up in test_os_exec.js? I am just trying to make the use clear. Is everyone happy with the way I've left that? 

I appreciate the feedback so far, I don't expect anyone to take up all their time on this. I just know what any windows user or admin needs, and I'm trying to complete it.

coreyapowell committed:
cleaning up and updating

removed the majority of flags. Only left flags that seemed necessary. 
I had a combined js_os_pipe with several #ifdefs inside to separate the OSs. I later restored the non-WIN32 version so that it would not be out of order from the code before. The logic was that a) the export list is maintained in the order of the function calls. and b) the code compare was less of a mess to keep that chunk intact.  I realized there were useless #ifdefs and removed them. I did retest both WIN32 and GNU.

coreyapowell committed:
Update quickjs-libc.c

I think I got everything!
mostly comments and reordering of if statements. 
I'm not the only programmer in the world to put my then branches on the same line following an if statement. But I'm genX, I used 640x480 CRTs. It looks clearer, I do admit that. But this is definitely a 21st century practice, in the old days we couldn't afford the Tabs or the printer paper.

coreyapowell committed:
Update test_os_exec.js

nothing functional, just tweaking the comments.
I'm not happy leaving this undocumented function in the wild. and it's trivial! but there it is! it at least deserved 2 lines of description.  It is more concise now .... just like Bellard would have wanted ... tidy and such.

past-due commented on quickjs-libc.c:
```diff
@@ -3473,10 +3653,114 @@ static JSValue js_os_exec(JSContext *ctx, JSValueConst this_val,
+    int pid;
+    HANDLE ph;
+    int block = 0, ret; 
+    DWORD options = 0, flags = PROCESS_QUERY_INFORMATION | SYNCHRONIZE;
```
This probably should be:
```suggestion
    DWORD options = 0, flags = PROCESS_QUERY_LIMITED_INFORMATION | SYNCHRONIZE;
```

coreyapowell replied:
tested and works fine like that. I'll put that in. There's better chances of it not tripping security checkpoints in someone else's random section of code with LTD INFO.

coreyapowell replied:
apparently you close PID HNDS. Unlike FDs to HND casts.

coreyapowell committed:
Update quickjs-libc.c

PROCESS_QUERY_LIMITED_INFORMATION

coreyapowell committed:
Update quickjs-libc.c

CloseHandle after OpenProcess. 
The documentation wasn't so clear, but this works as it should, and I read that it wastes resources not to. Don't close HNDs to FDs, let them get closed by the user. But close HNDs from PIDs.
Besides, this is a kill and an exit return value. It's stopped!

coreyapowell commented:
If you guys want to take some time reviewing this, I'm trying to pass these handles down pipes through stdin see if the child permissions work.

past-due commented:
> If you guys want to take some time reviewing this, I'm trying to pass these handles down pipes through stdin see if the child permissions work.

You may want to take a look at: https://learn.microsoft.com/en-us/windows/win32/sysinfo/handle-inheritance

coreyapowell committed:
Update test_os_exec.js

recombined the functions
tested against:
(linux) true
exit 2
bad command -> 127
good command ->0
watchpid blocking
environment
inherited pipe
waitpid blocking
watchpid nonblocking pid running
kill
watchpid nonblocking pid ended
waitpid WNOHANG

I have uncovered that linux echoes to the stdout and I am going to revert my flags for win32, but that change isn't tested yet in this file.

coreyapowell committed:
Update test_os_exec.js

don't sleep on blocking, moved comment too

coreyapowell committed:
Update quickjs-libc.c

"shell commands come through"

split the os_exec function back in half between the 2 versions.
fixed the way handles inherit and text buffers FROM SHELLS! text does not buffer from system commands, especially bad ones. Error messages are not the same. verifying 7 permutations are the same between the two OSs

coreyapowell committed:
Update test_os_exec.js

7 permutations of direct commands and shell commands firing error codes and text through console.
shells inherit stdio unless you pass a pipe. 
this is exec, bad programs don't execute. echo doesn't execute!

there's text output now. it may be to set a flag to disable text output for testing, or else execute a script and verify the pipe text.

coreyapowell committed:
Update test_os_exec.js

expanded 2 more functions and improved the test on file and inherit. nothing is  yet tested on cwd.

coreyapowell committed:
Update quickjs-libc.c

added inherit flag
it became more obvious that stdio should default to silent. set the inherit flag or pass one pipe, and the child will inherit the stdio of the parent (except the pipe you set). I believe this includes stdin!

coreyapowell commented:
silent by default, set inherit: true to pass stdio to the child, seems like the right thing to do.
linux version modified to do the same. I have a backup if that's not what you want.

coreyapowell committed:
Add files via upload

test_os_exec_child.js - executes the 8 basic permutative shell tests and verifies that only inherited handle text passes through stdout and stderr.
test_os_exec.js - now builds the test folder path and executes test_os_exec_child.js; verifies that stderr sees 'bad' and stdout only sees the proper echo string.
tests.conf will need update for all of this.

It comes to my attention there is no return value when doing asynchronous shell commands.

coreyapowell committed:
Update tests.conf

test_os_exec_child.js added to exclusions. there's no other way to silently verify stdout and stderr inheritance.

coreyapowell committed:
Update test_os_exec.js

amended child execution to test cwd parameter of os.exec

coreyapowell commented:
updated scripts to work with test262. Verifying 8 permutations is offloaded to test_os_exec_child.js. this verifies: 
1) inherited stdio, echo text passes through stdout,
2) 'bad' shell text passes through stderr.
3) using the cwd parameter to locate the child script.
using the word 'child' instead of 'slave', I've been programming hardware for 30 years and the political landscape has changed.

coreyapowell committed:
cwd tested from getcwd

I was building the script path, but I needed to test cwd. 
test262 is usually called from the make folder. 
manually I'm going into tests/ and calling from there so that assert.js is there.
now I look and see if 'tests' is the end of the path, if not I insert it, and I use 'cwd: dir,' 
If someone executes test262 from a build folder, I don't know how they get assert.js, but this will then break.

coreyapowell committed:
should work with test262

I need the cwd, and it needs testing. 
detects if you're in 'build'
detects if you're in 'scripts'
otherwise adds 'scripts'
then it builds the cwd to use and test.

coreyapowell committed:
removed test_os_exec

test_os_exec moved to other js

coreyapowell commented:
I need some eyes on this in case it won't work for you. 
I have to test cwd on os_exec, so i call os.getcwd(). But run-test262 will likely run from build or from the root project directory, so I pop 'build', and I push 'tests', and I join. Then I execute the child script from there, testing the cwd parameter.

I assume everything is fine, I can run-test262 without complaint. 
But I personally don't make 5 full folders from github, I make 5 build folders. So there's only room for the Makefile defaults. I'll do one win32 version and a WSL version, but I manually run the test scripts otherwise.
I can't think of a better way, but someone with a different build has to set this the way they want.
